Some minor fixes

jdunkerley opened the pull request:
### Pull Request Description

- Allow `Map` to store a `Nothing` key (fixes `Vector.distinct` with a `Nothing`).
- Add `column_names` method to `Table` as a shorthand.
- Return data flow error when comparing with Nothing (not a Panic or a Polyglot exception).
- Allow milli and micro second for DateTime and Time Of Day

### Important Notes

- Added a load of tests for the various comparison operators to Numbers_Spec.

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [ ] If GUI codebase was changed: Enso GUI was tested when built using BOTH
        `./run ide build` and `./run ide watch`.

jdunkerley committed:
Fix issues with Nothing in string compare.
Allow Nothing as a key in a Map.
Add column_names short hand to Table.
Add support for a Map in rename_columns.

jdunkerley committed:
Test graceful failing of Nothing in Text compare.
Adjusted Map to cope with Nothing keys better.
Tests for Nothing key in Map.

jdunkerley committed:
Make Decimal.compare_to error rather than Panic.

jdunkerley committed:
Working on making compare_to not Panic but Error

jdunkerley committed:
Adjust compare_to so they return an Error on a mismatched type.

jdunkerley committed:
Arguments the wrong way round... (for a long time!)

radeusgd commented on distribution/lib/Standard/Database/0.0.0-dev/src/Data/Table.enso:
```diff
@@ -523,6 +525,12 @@ type Table
     columns : Vector Column
     columns self = self.internal_columns . map self.make_column
 
+    ## UNSTABLE
+
+       Returns the vector of column names contained in this table.
+    column_names : Vector Text
+    column_names self = self.internal_columns . map _.name
```
Is this used anywhere?

radeusgd commented on engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/bool/CompareToNode.java:
```diff
@@ -29,11 +29,11 @@ Atom doBoolean(Boolean self, Boolean that) {
-    var typeError = Context.get(this).getBuiltins().error().makeTypeError(that, bool, "that");
-    throw new PanicException(typeError, this);
+    var typeError = Context.get(this).getBuiltins().error().makeTypeError(bool, that, "that");
+    return DataflowError.withoutTrace(typeError, this);
```
Raising this error without any trace will make it harder to debug.
While I appreciate that in many cases we want to avoid adding traces to Dataflow errors for performance, if we are shifting towards using dataflow errors much more than panics, I think we need to revisit that decision.

Maybe we should have 2 classes of errors? Cheap indicator errors that can happen in hot computations that need to be without trace for performance, and the more expensive 'real errors' that used to most often used to be panics but were shifted to dataflow. For the latter 'class', I'd add traces so we can find out where the error has happened.

JaroslavTulach replied:
Producing stacktrace is expensive. However including `getSourceSection()` - e.g. location in the source should be cheap - if that's enough. The `DataflowError` has reference to `this` - it can use `this.getSourceSection()` later to construct where the error comes from and that shall be fast.

radeusgd replied:
Interesting point.
Still for the more crucial errors I'd consider a full stacktrace as it can give much more information.
Since getting just the source location is so cheap, maybe we should include this for _all_ dataflow errors? That could be quite helpful for debugging still.

jdunkerley replied:
When I tested the code the `getSourceSection` didn't appear to work - I think it would be a good addition to the Dataflow error so suggest we add a task to pick this up later.

jdunkerley committed:
Update all the number nodes to return data flow errors.

jdunkerley committed:
Add a load of tests.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/number/bigInteger/GreaterOrEqualNode.java:
```diff
@@ -6,21 +6,20 @@
 
   @Specialization
-  boolean doDouble(EnsoBigInteger self, double that) {
+  Object doDouble(EnsoBigInteger self, double that) {
```
Do these `@Specialization` methods need to change return type? Can't they continue to more specialized types like `boolean` or `Atom` even the `execute` method returns `Object`?

JaroslavTulach reviewed (approved):
Please verify the effect on [benchmarks](https://github.com/enso-org/enso/actions/workflows/benchmark.yml). I think it should be OK, but one can never be sure. Run it for your branch and compare with `develop` ones.

Consider keeping the more specialized types in the `@Specialize` methods, if possible.

jdunkerley committed:
Milli and Micro second. Update the tests.

jdunkerley committed:
Fix date time tests.
Make specialisations return type more specific.

jdunkerley committed:
Changelog

jdunkerley committed:
Return types on Boolean.compare_to.
Fix Vector.distinct so returns error on mixed types.
Test for Table.rename_columns by Map.
Test for Table.column_names.

jdunkerley committed:
Fix Array.Sort.
Add assert to check BigInteger doesn't fit in Long.

hubertp committed:
Cleanup dataflow error propagation

Rather than going through instance checks, introduced a separate node
that uses specializations for that.
In the process, discovered that ComparatorNode was never used or covered
on our test suite. If it did, we would notice that we were comparing
Atoms with AtomConstructors, meaning that we would always throw a panic
when sort is given a comparator. It made more sense to remove that
specialization.
